### ISSUE TITLE
Ignore PCH-related command-line args

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#248]
Some projects compile with flags to compile/use pre-compiled headers.

This doesn't make much sense for clangd (conceivably we could build our own cache of them, but in practice the per-file preamble is probably enough).

This is particularly hilarious when we try to consume PCHes from a different version of clang... or from MSVC...

So we should strip such flags or mangle the invocations.

ccls has a filter on args so users can do e.g. ```
 "ccls.clang.excludeArgs": [
    "/YcC:*",
    "/YuC:*",
    "/FIC:*",
    "/FpC:*"
  ],
```

Maybe it's nice to have this user-servicable so they can work around future issues between releases, but we should fix the PCH flags by default.

**Comment by abique** [comment]
Good idea, as clangd also fails with gcc's pre-compiled headers..

**Comment by MarkZ3** [comment]
would it be a good idea to have the option to use the precompiled headers when they are from the same Clang version? Without precompiled header, opening a file the first time can be very slow.

**Comment by kadircet** [comment]
yes that would be nice. especially after finishing our preamble patching stuff, which will enable making use of stale preambles.

**Comment by sam-mccall** [comment]
> when they are from the same Clang version

Do you know of scenarios where this is likely to be true in practice?
(Serious question - I know Apple bundles a clangd that's version-locked with their toolchain, but not really aware of other cases.)

**Comment by MarkZ3** [comment]
For example, using the clang-cl and clangd from the same "LLVM" installation on Windows. Some distros could also package clang and clangd with matching versions. Projects also can bundle their own clang and clangd that are compatible (my case). Unless there is a smart way to detect compatibility by reading the pch, maybe just an option (flag) for clangd would be OK, the default would be not to use pch since it's the safest setting although possibly not the most optimal.

**Comment by bstaletic** [comment]
> would it be a good idea to have the option to use the precompiled headers when they are from the same Clang version?

Even this caused problems for ycmd, with libclang. If a user generates pch's with `/usr/bin/clang-10` ycmd's `third_party/clangd/output/bin/clangd` (same version) won't be able to consume it.

Reference: https://github.com/ycm-core/ycmd/issues/892

This is talking about libclang, but I tried it with clangd too. Looking at clangd's log, first clangd complained about a mismatching `__PIC__` value. Then `__PIE__`. Then complained about something else and I just gave up.
However, making ycmd use the system libclang/clangd made PCH's work.

Bottom line is - even matching the version isn't a guarantee that PCH's will work. The compiler needs to exactly match.

**Comment by MarkZ3** [comment]
> The compiler needs to exactly match.

That's my main scenario so I'll look at adding an option, with enough information in the description to prevent confusion.
